### PR TITLE
Improve performance of `is_arrays_compatible`

### PR DESCRIPTION
**Pull request by niboshi**
For single `numpy.ndarray` argument: 1101 ns -> 654 ns per call.
For three `numpy.ndarray` arguments: 1299 ns -> 820 ns per call.

The performance of backprop (#6263) slightly improves from 19.1 ms -> 18.4 ms.

### CODE REVIEW

File: chainer/__init__.py [hunk 1]
```diff
@@ -160,25 +160,60 @@ def get_cpu_array_types():
...
+            has_ideep = has_ideep or isinstance(a, backends.intel64.mdarray)
+            if has_chainerx and has_ideep:
+                # chainerx arrays and intel64.mdarray are not compatible.
+                return False
+
+        if not has_chainerx:
+            # `types` is only initialized when needed.
+            # (i.e. at the first occurrence of non-None, non-ChainerX inputs)
+            if types is None:
+                if isinstance(a, backends.cuda.ndarray):
```
Comment (okuta) [inline chainer/__init__.py]: How about following code?
```
types = backends.cuda.ndarray
if not isinstance(a, types):
    types = get_cpu_array_types()
    if not isinstance(a, types):
        return False
```

File: chainer/__init__.py [hunk 2]
```diff
@@ -160,25 +160,60 @@ def get_cpu_array_types():
...
+                        # TODO(niboshi): Perhaps better to make it an error but
+                        # in order to do that we won't be able to early-return
+                        # False from within the loop.
+                        return False
+
+            elif not isinstance(a, types):
+                # cpu/gpu arrays are mixed.
+                has_mixed_cpu_gpu = True
+
+    if not has_chainerx and has_mixed_cpu_gpu:
```
Comment (okuta) [inline chainer/__init__.py]:
```suggestion
    return has_chainerx or not has_mixed_cpu_gpu
```

Comment (niboshi) [reply chainer/__init__.py]: Why do you want this fix?
I want to leave the comment in the current code, and the code is more straightforward considering the comment. This condition can be considered as a kind of early-return case.

File: chainer/__init__.py (hunk 1 above)
Comment (niboshi) [reply chainer/__init__.py]: Fixed.

**Comment by toslunar** [comment]
The PR will conflict with #7520.

**Pull request closed without merging**